### PR TITLE
feat: create annotate only from diff coverage

### PR DESCRIPTION
**Pull request by MartinBernstorff**
Tried my hand here! A few challenges remain, mainly that I'm not used to Jinja and doing web-mocking in Python, so unsure how to fix those test errors :-)

### CODE REVIEW

**Comment by MartinBernstorff** [comment]
A few type issues were caught by pyright that mypy seems to have missed - I've chosen to ignore them for now, but do recommend pyright over mypy 👍 

https://github.com/microsoft/pyright/blob/main/docs/mypy-comparison.md

**Review by kieferro — COMMENTED** [review]
Thank you for the PR :) That looks pretty good

> I'm not used to Jinja and doing web-mocking in Python, so unsure how to fix those test errors :-)

One of the tests that is not successful is simply because the output [here](https://github.com/MartinBernstorff/python-coverage-comment-action/blob/b7c13c9668c6953e3d7d251f001762c610f1a396/tests/integration/test_main.py#L264) has changed because line 6 is not part of the `DiffCoverage`, but part of the original codebase. So that actually proves that the changes work 😄

File: tests/conftest.py [hunk 1]
```diff
@@ -281,7 +282,7 @@ def diff_coverage_obj():
         total_percent_covered=decimal.Decimal("0.8"),
         num_changed_lines=39,
         files={
-            "codebase/code.py": coverage_module.FileDiffCoverage(
+            Path("codebase/code.py"): coverage_module.FileDiffCoverage(
```
[the remark this answers is not in the document]
Comment (MartinBernstorff) [reply tests/conftest.py]: I've modified the annotation to `str` for now :-) 

@ewjoachim

**Comment by MartinBernstorff** [comment]
Thanks for an excellent review experience, @kieferro! I'm still not quite sure how `tests/integration/test_main.py::test_action__pull_request__annotations` creates `foo.py`, and what git history it uses.

My understanding is that the integration test environment is setup in `integration_env`. I can see it create files `A` and `B`, then switch to another branch and create file `C`. However, I don't know when it creates `foo.py`. 

`get_diff_coverage_info` used `origin/main` as its comparator to get diff coverage. I've added an option to not use origin, but we get no diff from `foo.py`. When I know where `foo.py` is created, I might be able to debug it :-)

Let me know what you guys think, and feel free to commit to it as well 👍

File: tests/unit/test_annotations.py [hunk 2]
```diff
@@ -12,20 +14,14 @@ def test_annotations(coverage_obj, capsys):
...
-::notice file=codebase/main.py,line=3::This line has no coverage
-::notice file=codebase/main.py,line=7::This line has no coverage
-::notice file=codebase/main.py,line=13::This line has no coverage
-::notice file=codebase/main.py,line=21::This line has no coverage
-::notice file=codebase/main.py,line=123::This line has no coverage
-::notice file=codebase/caller.py,line=13::This line has no coverage
-::notice file=codebase/caller.py,line=21::This line has no coverage
-::notice file=codebase/caller.py,line=212::This line has no coverage
+::notice file=codebase/code.py,line=7::This line has no coverage
+::notice file=codebase/code.py,line=9::This line has no coverage
```
[the remark this answers is not in the document]
Comment (kieferro) [reply tests/unit/test_annotations.py]: > Is there any reason that the n_lines>2 case is different from the n_lines==2 case in the other test in the file?

This test is mainly for testing a bit more extensively (more files, more missing lines, more edge cases). I have committed a suggestion how I would imagine this.

**Comment by kieferro** [comment]
> I can see it create files `A` and `B`, then switch to another branch and create file `C`.

These are actually not files, just variables created in `foo.py`. If you look closely [here](https://github.com/MartinBernstorff/python-coverage-comment-action/blob/ce3a9157d67a8029e9737d87fb5fad7764e5a817/tests/integration/test_main.py#L32) you can see that `file_path` is just a fixture and "A", "B" and "C" as parameters are the variables and not the filenames. That means the file `foo.py` will look like this in the end:
```python
if os.environ.get("A"):
 1
if os.environ.get("B"):
 2
if os.environ.get("C"):
 3
```

And the creation of `foo.py` happens indirectly [here](https://github.com/MartinBernstorff/python-coverage-comment-action/blob/ce3a9157d67a8029e9737d87fb5fad7764e5a817/tests/integration/test_main.py#L27), by using this fixture to write the code with the variables into it.


I hope that was understandable. If not, feel free to ask, I'm always happy to support with explanations (or code) 😃 


> Let me know what you guys think, and feel free to commit to it as well 👍

Thanks :+1:, I have added two commits with small changes.

**Comment by ewjoachim** [comment]
Good ! I have some changes I drafted but haven't posted yet on changing the integration fixture so that there is some diff coverage lines missing in the general case (changes the 77% figure in quite a few different test. I guess the best course of action would be to merge as is (when tests pass) and I'll do a follow-up PR with new tests, so as not to delay this too much.

**Comment by ME-ON1** [comment]
Screenshot 2023-06-16 at 5 32 04 PM

Hey i think this has broken the `--compare-branch` feature

**Comment by kieferro** [comment]
@MartinBernstorff Unfortunately I had to revert this PR as a quickfix, because it caused the action to not work properly (#192).
That this happened is in no way your fault. We as reviewers are responsible and we should have checked this more carefully before we merge. Nevertheless we would like to have this feature and your contribution in the codebase. (If you want) you are welcome to reopen the same PR from your branch and we'll review again in more detail and hopefully without missing anything.
